tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0009_prefer_course_key.py

```diff
@@ -0,0 +1,213 @@
+            JSON_VALUE(event_str, '$.result.score.scaled') as scaled_score,
+            if(
+                verb_id = 'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                cast(JSON_VALUE(event_str, '$.result.success') as Bool),
+                false
+            ) as success,
+            JSON_VALUE(event_str, '$.object.definition.interactionType') as interaction_type,
+            if(
+                verb_id = 'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                cast(JSON_VALUE(event_str, '$.object.definition.extensions."http://id.tincanapi.com/extension/attempt-id"') as Int16),
+                0
+            ) as attempts
+        FROM
+            {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+        WHERE
+            verb_id in (
+                'https://w3id.org/xapi/acrossx/verbs/evaluated',
+                'http://adlnet.gov/expapi/verbs/passed',
+                'http://adlnet.gov/expapi/verbs/asked'
+            );
+        """,
+    "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_TRANSFORM_MV }}": """
+        CREATE MATERIALIZED VIEW IF NOT EXISTS {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_TRANSFORM_MV }}
+        TO {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_EVENTS_TABLE }} AS
+        SELECT
+            event_id,
+            emission_time,
+            actor_id,
+            object_id,
+            course_id,
+            org,
+            verb_id,
+            JSON_VALUE(event_str, '$.object.definition.type') AS object_type,
+            -- clicking a link and selecting a module outline have no starting-position field
+            if (
+                object_type in (
+                    'http://adlnet.gov/expapi/activities/link',
+                    'http://adlnet.gov/expapi/activities/module'
+                ),
+                0,
+                cast(JSON_VALUE(
+                    event_str,
+                    '$.context.extensions."http://id.tincanapi.com/extension/starting-position"'
+                ) as Int16)
+            ) AS starting_position,
+            JSON_VALUE(
+                event_str,
+                '$.context.extensions."http://id.tincanapi.com/extension/ending-point"'
+            ) AS ending_point
+        FROM
+            {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+        WHERE verb_id IN (
+            'https://w3id.org/xapi/dod-isd/verbs/navigated'
+        );
+        """,
+}
+
+
+def upgrade():
+    # drop the tables and MVs and create new ones, with course_id
+    # replaced by course_key
+    for name, ddl in TABLES.items():
+        op.execute(f"DROP TABLE IF EXISTS {name}")
+        op.execute(ddl.replace("course_id", "course_key"))
+
+    for name, query in MATERIALIZED_VIEWS.items():
+        op.execute(f"DROP TABLE {name}")
+        op.execute(
+            query.replace(
+                "course_id", "splitByString('/', course_id)[-1] AS course_key"
+            )
+        )
+
```

Comment:
We'll also want to populate the new tables with the results of the MV queries, sorry I wasn't clear what I was asking for there. I meant we could take the select statements and re-use them after create to do `insert into ... select from...`.

Reply:
Ah got it, I'll change this so we create new tables, populate the data, then rename the existing tables before dropping them.